src/main/serialport.ts

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
+ */
+
+import type { AutoDetectTypes, SetOptions } from '@serialport/bindings-cpp';
+import { Renderer, WebContents } from 'electron';
+import { SERIALPORT_CHANNEL } from 'pc-nrfconnect-shared/main';
+import { SerialPort, SerialPortOpenOptions } from 'serialport';
+
+export type Renderer = Pick<WebContents, 'on' | 'send' | 'id'>;
+
+/**
+ *  serialPorts: {
+
+        '/dev/Robot': {
+            renderers: Renderer[],
+            serialPort: SerialPort,
+        },
+
+        '/dev/Rowboat': {
+            renderers: Renderer[],
+            serialPort: SerialPort,
+        }
+    }
+*/
+type OpenPort = {
+    serialPort: SerialPort;
+    renderers: Renderer[];
+    settingsLocked: boolean;
+    opening: boolean;
+};
+
+const serialPorts = new Map<string, OpenPort>();
+
+export const openOrAdd = async (
+    sender: Renderer,
+    options: SerialPortOpenOptions<AutoDetectTypes>,
+    overwrite: boolean
+) => {
+    const { path } = options;
+    const existingPort = serialPorts.get(path);
+
+    if (existingPort) {
+        // Port is already opened
+
+        if (existingPort.opening) {
+            throw new Error('PORT_IS_ALREADY_BEING_OPENED');
+        }
+
+        const renderers = existingPort.renderers;
+        const alreadyInList = renderers.find(
+            existing => sender.id === existing.id
+        );
+
+        if (!alreadyInList) {
+            let isDifferentSettings;
+            if (existingPort.serialPort) {
+                const currentOptions = existingPort.serialPort.settings;
+                Object.entries(options).forEach(([key, value]) => {
+                    if (
+                        !(key in currentOptions) ||
+                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                        (currentOptions as any)[key] !== value
+                    ) {
+                        isDifferentSettings = true;
+                    }
+                });
```

Comment:
consider using every e.g.

```suggestion
                Object.entries(options).every(([key, value])=> {
                    if (
                        !(key in currentOptions) ||
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        (currentOptions as any)[key] !== value
                    ) {
                        isDifferentSettings = true;
                        return false;
                    }

                    return true;
                });
```

and return false when isDifferentSettings  = true to exit the loop and we don't care if we have more conflicts one is enough